HW2/P2/parallel_vector.pyx

```diff
@@ -72,19 +72,44 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
                              int repeat):
    cdef:
        int idx, r
+       int maxV, minV
        omp_lock_t *locks = get_N_locks(counts.shape[0])
 
    ##########
    # Your code here
    # Use parallel.prange() and a lock for each element of counts to parallelize
    # data movement.  Be sure to avoid deadlock, and double-locking.
+
    ##########
    with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+           for r in prange(repeat, num_threads = 4):
+               for idx in range(src.shape[0]):
+                   # No change to counts
+                   if src[idx] == dest[idx]:
+                       continue
+
+                   # Impose locking order A then B where A < B
+                   minV = src[idx] if src[idx] < dest[idx] else dest[idx]
```

Comment:
you store the values of minV and maxV here, but what if the true evaluation of src[idx] > dest[idx] changes between now and when the lock is acquired (which may happen if any of the values change!)